protocols/autonat/src/behaviour.rs

```diff
@@ -0,0 +1,853 @@
+            ConnectedPoint::Listener { .. } => {
+                if self.confidence == self.config.confidence_max {
+                    if self.nat_status.is_public() {
+                        self.schedule_next_probe(self.config.refresh_interval * 2);
+                    } else {
+                        self.schedule_next_probe(self.config.refresh_interval / 5);
+                    }
+                }
+            }
+        }
+    }
+
+    fn inject_dial_failure(
+        &mut self,
+        peer_id: Option<PeerId>,
+        handler: Self::ProtocolsHandler,
+        error: &DialError,
+    ) {
+        self.inner.inject_dial_failure(peer_id, handler, error);
+        if let Some((_, channel)) = peer_id.and_then(|p| self.ongoing_inbound.remove(&p)) {
+            log::debug!(
+                "Dial-back to peer {} failed with error {:?}.",
+                peer_id.unwrap(),
+                error
+            );
+
+            let response = DialResponse {
+                result: Err(ResponseError::DialError),
+                status_text: Some("dial failed".to_string()),
+            };
+            let _ = self.inner.send_response(channel, response);
+        }
+    }
+
+    fn inject_disconnected(&mut self, peer: &PeerId) {
+        self.inner.inject_disconnected(peer);
+        self.connected.remove(peer);
+    }
+
+    fn inject_address_change(
+        &mut self,
+        peer: &PeerId,
+        conn: &ConnectionId,
+        old: &ConnectedPoint,
+        new: &ConnectedPoint,
+    ) {
+        self.inner.inject_address_change(peer, conn, old, new);
+
+        self.connected
+            .insert(*peer, new.get_remote_address().clone());
+    }
+
+    fn inject_new_listen_addr(&mut self, id: ListenerId, addr: &Multiaddr) {
+        self.inner.inject_new_listen_addr(id, addr);
+
+        // New address could be publicly reachable, trigger retry.
+        if !self.nat_status.is_public() {
+            if self.confidence > 0 {
+                self.confidence -= 1;
+            }
+            self.schedule_next_probe(self.config.retry_interval);
+        }
+    }
+
+    fn inject_expired_listen_addr(&mut self, id: ListenerId, addr: &Multiaddr) {
+        self.inner.inject_expired_listen_addr(id, addr);
+        if let Some(public_address) = self.public_address() {
+            if public_address == addr {
+                self.confidence = 0;
+                self.nat_status = NatStatus::Unknown;
+                self.schedule_probe.reset(Duration::ZERO);
+            }
+        }
+    }
+
+    fn inject_new_external_addr(&mut self, addr: &Multiaddr) {
+        self.inner.inject_new_external_addr(addr);
+
+        // New address could be publicly reachable, trigger retry.
+        if !self.nat_status.is_public() {
+            if self.confidence > 0 {
+                self.confidence -= 1;
+            }
+            self.schedule_next_probe(self.config.retry_interval);
```

Comment:
In general, `NetworkBehaviour::poll` is guaranteed to be called after `NetworkBehaviour::inject_xxx` is called unless directly triggered through synchronous methods on `Swarm`, e.g. `Swarm::add_external_address`. As you noted this is a problem here as `NetworkBehaviour::inject_new_external_addr` potentially resets the timer.

I am not sure how to tackle this issue. As you mentioned in the main thread, keeping the previous `Waker` around within `Behaviour` seems rather unclean.

Given that a call to `NetworkBehaviour::poll` is triggered through all kinds of external mechanisms (e.g. a network event, a `Delay` firing in another `NetworkBehaviour`, ...) and thus the `Waker` is properly registered with the `Delay` of `self.schedule_next_probe`, I wonder how often this happens in the wild. In other words I wonder how often a user would call a method on `Swarm` triggering `self.schedule_next_probe` to reset followed by a period with zero external inputs (e.g. network events, ...).

The above said, depending on external mechanisms (e.g. network events) is less than ideal and likely error prone.

Are you able to trigger this error in unit tests?

Reply:
I tried to test this directly on the `autonat::Behaviour` and on a Swarm, but in both cases the test surprisingly passes: b3670b7.
I honestly don't understand why the waker is wakened again in this case, but not if I reset the delay within `Behaviour::poll`. Any ideas?

Reply:
So if I understand you correctly, the fact that we **actively** call again `Behaviour::poll` / `Swarm::poll` after `inject_new_external_addr` registers the waker again?
If that is the case, doesn't this also solves the above issue of the user manually triggering Swarm methods like `inject_new_external_addr`, that reset the timer? Because in order to call such a method, a user would need to stop polling the swarm, and then actively start polling it again.

Reply:
> So if I understand you correctly, the fact that we **actively** call again `Behaviour::poll` / `Swarm::poll` after `inject_new_external_addr` solves it that the waker is registered again?

Correct.

> If that is the case, doesn't this also solves the above issue of the user manually triggering Swarm methods like `inject_new_external_addr`, that reset the timer? Because in order to call such a method, a user would need to stop polling the swarm, call the method, and then actively start polling it again, hence also registering the new waker.

In the ideal case yes, though the user could as well do something like:

``` rust
loop {
  let is_pending = my_swarm.poll_next(cx).is_pending();
  my_swarm.add_external_address(some_address);
  if is_pending {
    return Poll::Pending;
  }
}

Reply:
Ah yes you are right, I didn't think about that. I am able to trigger this error with https://github.com/elenaf9/rust-libp2p/commit/ae95520449da3e807b9efef9b603a3883db43600, the test there never terminates.
Not sure if in practice anyone would implement something like this, but I have to admit it is not too far-fetched.
I don't know if this is a show-stopper or not. If we want to be 100% sure I can just remove the resetting of the timer from any method apart from `NetworkBehaviour::poll`. Worst case with this then would mean that we have to wait a whole `refresh_period` until a next probe.

Reply:
I am leaning towards keeping it as is, betting on users not to write logic like https://github.com/elenaf9/rust-libp2p/commit/ae95520449da3e807b9efef9b603a3883db43600. I find it intuitive to have to `poll` something after injecting an event (i.e. calling `add_external_address`).

Reply:
Agreed.